### PR TITLE
Editorial: Move NumberFormat objects to DateTimeFormat constructor an…

### LINKED ISSUE #563: Clarification on PartitionDateTimePattern: Constructing NumberFormat instances
Ref: https://402.ecma-international.org/7.0/#sec-partitiondatetimepattern

```ts
Perform ! CreateDataPropertyOrThrow(nfOptions, "useGrouping", false).
Let nf be ? Construct(%NumberFormat%, « locale, nfOptions »).
Let nf2Options be ObjectCreate(null).
Perform ! CreateDataPropertyOrThrow(nf2Options, "minimumIntegerDigits", 2).
Perform ! CreateDataPropertyOrThrow(nf2Options, "useGrouping", false).
Let nf2 be ? Construct(%NumberFormat%, « locale, nf2Options »).
Let tm be ToLocalTime(x, dateTimeFormat.[[Calendar]], dateTimeFormat.[[TimeZone]]).
```

The steps in `PartitionDateTimePattern` include constructing two separate `NumberFormat` instances. (step 6 and 10)

This operation is called when formatting every single date and it can be really slow.

Caching the two `NumberFormat` instances by locale instead of constructing them every time can increase the performance of formatting dates [anywhere from 20x to 100x](https://github.com/formatjs/formatjs/pull/2827). This is especially important for lower end devices, or for jitless runtimes.

Is there any recommendation on how to handle this? Is caching appropriate here?

How do other implementations handle this?

### PR DESCRIPTION
**Pull request by longlho**
…d add new internal slots

fix #563

### CODE REVIEW

File: spec/datetimeformat.html
Comment (ryzokuken) [inline spec/datetimeformat.html:359]: Both of these lines can be removed.

File: spec/datetimeformat.html
Comment (ryzokuken) [inline spec/datetimeformat.html:360]: Nit: What do you think about renaming this to `_nf1_` everywhere?

File: spec/datetimeformat.html
Comment (ryzokuken) [inline spec/datetimeformat.html:191]: There is no `_locale_` here, you should use either `_r_.[[locale]]` or `_dateTimeFormat_.[[Locale]]`, preferably the latter IMO.

File: spec/datetimeformat.html
Comment (ryzokuken) [inline spec/datetimeformat.html:189]: Nit: what do you think about renaming this to `_nf1Options_`?

File: spec/datetimeformat.html
Comment (ryzokuken) [inline spec/datetimeformat.html:195]: Same as above for `_locale_`.

File: spec/datetimeformat.html
Comment (ryzokuken) [inline spec/datetimeformat.html:201]: Same as above for `_locale_`.

File: spec/datetimeformat.html
Comment (ryzokuken) [inline spec/datetimeformat.html:1261]: Should there be more information here about these three? Also since `[[NumberFormat3]]` may or may not be set based on `fractionalSecondDigits`, I suppose we should mention that too?

**Review by ryzokuken — CHANGES_REQUESTED** [review]
A couple of comments, thanks for the PR!

**Comment by longlho** [comment]
@ryzokuken done. Thanks a lot for the review!

**Comment by sffc** [comment]
CC @FrankYFTang -- can you review this proposed change from @longlho and comment on whether it would impact implementations like V8?

**Comment by ryzokuken** [comment]
@sffc I think @anba mentioned that it wouldn't affect ICU-based implementations.

**Comment by FrankYFTang** [comment]
> CC @FrankYFTang -- can you review this proposed change from @longlho and comment on whether it would impact implementations like V8?

I believe this is editorial and has no implementation impact to v8.

**Review by FrankYFTang — APPROVED** [review]
LGTM

**Review by ryzokuken — APPROVED** [review]
Looks perfect now, thanks @longlho!

**Comment by ryzokuken** [comment]
@gibson042 @leobalter can we get one editor +1 and merge this? :D

**Comment by dlebedynskyi** [comment]
Any update on merging this pull? Looks like fairly important optimization

File: spec/datetimeformat.html
Comment (anba) [inline spec/datetimeformat.html:1259]: Slight rewording to match the other slot descriptions:
- Prefix with "Intl." match <https://tc39.es/ecma402/#sec-properties-of-intl-relativetimeformat-instances>. (Similar changes should be applied to the other two internal slot descriptions.)
- "It is only used when" to match <https://tc39.es/ecma402/#sec-properties-of-intl-numberformat-instances> and <https://tc39.es/ecma402/#sec-properties-of-intl-datetimeformat-instances>. (DateTimeFormat repeats the slot name [[HourCycle]], but I think it's okay to avoid that repetition.)

```suggestion
      <li>[[NumberFormat3]] is an Intl.NumberFormat object used for formatting fractional second digits. It is only used when [[FractionalSecondDigits]] is not *undefined*.</li>
```

File: spec/datetimeformat.html
Comment (anba) [inline spec/datetimeformat.html:390]: We could probably just directly access the internal slot instead of using variables at this point. 
```suggestion
              1. Let _fv_ be FormatNumeric(_dateTimeFormat_.[[NumberFormat1]], _v_).
```

File: spec/datetimeformat.html
Comment (anba) [inline spec/datetimeformat.html:715]: Pre-existing issue: Can you sneak in the three missing internal slots while modifying this line? Thanks! :smile: 

```suggestion
        1. Let _dateTimeFormat_ be ? OrdinaryCreateFromConstructor(_newTarget_, *"%DateTimeFormat.prototype%"*, &laquo; [[InitializedDateTimeFormat]], [[Locale]], [[Calendar]], [[NumberingSystem]], [[TimeZone]], [[Weekday]], [[Era]], [[Year]], [[Month]], [[Day]], [[DayPeriod]], [[Hour]], [[Minute]], [[Second]], [[FractionalSecondDigits]], [[TimeZoneName]], [[HourCycle]], [[DateStyle]], [[TimeStyle]], [[Pattern]], [[RangePatterns]], [[BoundFormat]], [[NumberFormat1]], [[NumberFormat2]], [[NumberFormat3]] &raquo;).
```

File: spec/datetimeformat.html
Comment (anba) [inline spec/datetimeformat.html:191]: I think I'd prefer to use `dateTimeFormat.[[Locale]]` here, because that matches the surrounding code a bit better.
```suggestion
        1. Set _dateTimeFormat_.[[NumberFormat1]] to ? Construct(%NumberFormat%, &laquo; _dateTimeFormat_.[[Locale]], _nf1Options_ &raquo;).
```

File: spec/datetimeformat.html
Comment (anba) [inline spec/datetimeformat.html:1257]: Maybe these should have more descriptive names?

**Review by anba — CHANGES_REQUESTED** [review]

**Comment by longlho** [comment]
@anba care for another look? thanks!

**Comment by ryzokuken** [comment]
Ping @anba PTAL.

**Comment by longlho** [comment]
@ryzokuken any update on this?

**Comment by ryzokuken** [comment]
@longlho I'd been waiting for @anba to give their approval before merging but I suppose I could go through their comments later today and see if their concerns have been addressed and merge if they are busy right now.

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:191]: If these uses of Construct could actually throw, then the PR would be normative. But I don't see how they can.
```suggestion
        1. Set _dateTimeFormat_.[[NumericNumberFormat]] to ! Construct(%NumberFormat%, &laquo; _dateTimeFormat_.[[Locale]], _nf1Options_ &raquo;).
```

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:195]:
```suggestion
        1. Set _dateTimeFormat_.[[2DigitNumberFormat]] to ! Construct(%NumberFormat%, &laquo; _r_.[[locale]], _nf2Options_ &raquo;).
```

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:201]:
```suggestion
          1. Set _dateTimeFormat_.[[FractionalSecondNumberFormat]] to ! Construct(%NumberFormat%, &laquo; _r_.[[locale]], _nf3Options_ &raquo;).
```

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:365]: This branch should assert that _dateTimeFormat_.[[FractionalSecondNumberFormat]] has the expected kind of value, since assignment to that slot is conditional.
```suggestion
          1. Else if _p_ is equal to *"fractionalSecondDigits"*, then
            1. Let _nf_ be _dateTimeFormat_.[[FractionalSecondNumberFormat]].
            1. Assert: Type(_nf_) is Object and _nf_ has an [[InitializedNumberFormat]] internal slot.
```

**Review by gibson042 — APPROVED** [review]

**Comment by longlho** [comment]
thanks a lot for the fixes @gibson042 !

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:195]: Consistency nit:
```suggestion
        1. Set _dateTimeFormat_.[[2DigitNumberFormat]] to ! Construct(%NumberFormat%, &laquo; _dateTimeFormat_.[[Locale]], _nf2Options_ &raquo;).
```

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:201]: Consistency nit:
```suggestion
          1. Set _dateTimeFormat_.[[FractionalSecondNumberFormat]] to ! Construct(%NumberFormat%, &laquo; _dateTimeFormat_.[[Locale]], _nf3Options_ &raquo;).
```

File: spec/datetimeformat.html
Comment (gibson042) [inline spec/datetimeformat.html:370]:
```suggestion
            1. Let _fv_ be FormatNumeric(_nf_, _v_).
```

**Review by gibson042 — APPROVED** [review]
A few more suggestions, but LGTM... thanks @longlho!

**Comment by longlho** [comment]
gentle ping @anba :)

**Comment by longlho** [comment]
ping @ryzokuken

**Comment by longlho** [comment]
@ryzokuken do you know if this is still mergeable? happy to rebase if that's the case

**Comment by ryzokuken** [comment]
Sorry for the delay, @longlho. It looks all good from my side, but looks like Andre still couldn't review it. If you rebase it, perhaps we could merge it and file an issue to address their concerns in a follow-up?

**Comment by sffc** [comment]
@anba @longlho @ryzokuken Who needs to take the next action on this PR?

**Comment by anba** [comment]
Hmm, this change may no longer be possible, because the current Temporal draft spec changes `Intl.DateTimeFormat` to be able to format Temporal objects with different resolved patterns. For example when formatting `Temporal.PlainTime` and `Temporal.PlainDateTime` with a single `Intl.DateTimeFormat`, the resolved pattern for `Temporal.PlainTime` can have a different amount of `fractionalSecondDigits` when compared to the resolved pattern for `Temporal.PlainDateTime`. Therefore it's not possible to store a single `[[FractionalSecondNumberFormat]]` internal slot.

**Comment by sffc** [comment]
Oh I see, yeah, I don't think we should do this. The spec does not forbid implementations from doing internal optimizations like this so long as the observable behavior is the same.

I'll close the PR since it has been idle for a long time and it's not clear if it is a good PR based on this and @anba's comment.

**Pull request closed without merging**
